File: GeometricObjects/DiskDetector.py
from RTree import *
from Vector import *
import numpy as np

import math




class DiskDetector:
    def __init__(self, R: float, center: Vector, normal: Vector, n=500) -> None:
        self.R = R
        self.center = center
        self.normal = normal
        self.surfacePoints: list[Vector] = self.getSurfacePoints(n)

    def getRotationAxisAndAngle(self, initial_normal: Vector, target_normal:Vector):
        """Compute the rotation axis and angle to align initial_normal with target_normal."""
        # Normalize the vectors
        initial_normal = initial_normal.norm()
        target_normal = target_normal.norm()
        #print(initial_normal.dot(target_normal) - 1)
        if (initial_normal.dot(target_normal) - 1)**2 < 1e-8:
            dot_product = initial_normal.dot(target_normal)
            if dot_product > 0:
                # No rotation needed
                return Vector(0, 0, 0), 0
            else:
                # 180-degree rotation around any axis perpendicular to the normals
                return Vector(1, 0, 0), np.pi

        # Compute rotation axis (it's perpendicular to the plane formed by initial_normal and target_normal)
        rotation_axis = initial_normal.cross(target_normal).norm()


        # If the rotation axis has a magnitude of 0, it means the initial and target normals are the same or opposite.

        # Compute rotation angle using dot product
        cos_angle = initial_normal.dot(target_normal)
        rotation_angle = np.arccos(cos_angle)

        return rotation_axis, rotation_angle

    def getSurfacePoints(self, n)-> list[Vector]:
        points = []

        # Calculate the side length of the squares in the grid
        side_length = 2 * self.R / np.sqrt(n)

        # Number of grid points along one dimension
        num_points = int(2 * self.R / side_length)
        opoints = []
        for i in range(num_points):
            for j in range(num_points):
                x = -self.R + i * side_length
                y = -self.R + j * side_length

                # Check if the point is inside the disk
                if x**2 + y**2 <= self.R**2:
                    point = Vector(x, y, 0)  # This is in the disk's local coordinates
                    #print("pt", point)
                    opoints.append(point)
                    # Rotate point based on disk's normal
                    rotation_axis, rotation_angle = self.getRotationAxisAndAngle(
                        Vector(0, 0, 1), self.normal
                    )
                    #print("angle, axis", rotation_angle, rotation_axis)
                    #print("1", point)
                    point = point.rotate(rotation_axis, rotation_angle)
                    #print("2", point)
                    # Translate to disk's position
                    point = point + self.center

                    points.append(point)

        return points
    
    def isVisible(self, rtree: RTree, source: Vector):
        for point in self.surfacePoints:
            dir = point - source
            hit = rtree.checkIntersections(source, dir)
            #print(hit)
            if(hit.tri is None or (hit.didHit and dir.find_t(source, point) < hit.t)):
                return True
        return False
    

    

File: rust/src/diskdetector.rs
use crate::rtree::RTree;
use crate::vector::Vector;
use std::f32::consts::PI;

pub struct DiskDetector {
    pub r: f32,
    pub center: Vector,
    pub normal: Vector,
    pub surface_points: Vec<Vector>,
}

impl DiskDetector {
    pub fn new(r: f32, center: Vector, normal: Vector, n: usize) -> Self {
        let surface_points = DiskDetector::get_surface_points(r, center, normal, n);
        DiskDetector {
            r,
            center,
            normal,
            surface_points,
        }
    }

    fn get_rotation_axis_and_angle(initial_normal: Vector, target_normal: Vector) -> (Vector, f32) {
        let initial_normal = initial_normal.norm();
        let target_normal = target_normal.norm();
        if (initial_normal.dot(target_normal) - 1.0).abs() < 1e-8 {
            let dot_product = initial_normal.dot(target_normal);
            if dot_product > 0.0 {
                return (
                    Vector {
                        x: 0.0,
                        y: 0.0,
                        z: 0.0,
                    },
                    0.0,
                );
            } else {
                return (
                    Vector {
                        x: 1.0,
                        y: 0.0,
                        z: 0.0,
                    },
                    PI,
                );
            }
        }
        let rotation_axis = initial_normal.cross(target_normal).norm();
        let cos_angle = initial_normal.dot(target_normal);
        let rotation_angle = cos_angle.acos();
        (rotation_axis, rotation_angle)
    }

    fn get_surface_points(r: f32, center: Vector, normal: Vector, n: usize) -> Vec<Vector> {
        let mut points = Vec::new();

        let side_length = 2.0 * r / (n as f32).sqrt();
        let num_points = (2.0 * r / side_length).round() as usize;

        for i in 0..num_points {
            for j in 0..num_points {
                let x = -r + i as f32 * side_length;
                let y = -r + j as f32 * side_length;

                if x.powi(2) + y.powi(2) <= r.powi(2) {
                    let mut point = Vector { x, y, z: 0.0 };
                    let (rotation_axis, rotation_angle) = Self::get_rotation_axis_and_angle(
                        Vector {
                            x: 0.0,
                            y: 0.0,
                            z: 1.0,
                        },
                        normal.clone(),
                    );
                    point = point.rotate(rotation_axis, rotation_angle);
                    point = point + center;
                    points.push(point);
                }
            }
        }

        points
    }

    pub fn is_visible(&self, rtree: &RTree, source: Vector) -> bool {
        for point in &self.surface_points {
            let dir = *point - source;
            if let Some(hit) = rtree.check_intersections(source.clone(), dir.clone()) {
                return true;
            }
        }
        false
    }
}
